src/ssl.c

```diff
@@ -5650,13 +5667,37 @@ int wolfSSL_Init(void)
     WOLFSSL_ENTER("wolfSSL_Init");
 
 #ifndef WOLFSSL_MUTEX_INITIALIZER
-    if (inits_count_mutex_valid == 0) {
-        if (wc_InitMutex(&inits_count_mutex) != 0) {
-            WOLFSSL_MSG("Bad Init Mutex count");
-            return BAD_MUTEX_E;
+    if (inits_count_mutex_valid <= 0) {
+    #ifdef WOLFSSL_CLEANUP_THREADSAFE
+        int current_inits_count_mutex_valid;
+        if (wolfSSL_Atomic_Int_FetchAdd(&inits_count_mutex_valid2, 1) != 0) {
+            /* We treat inits_count_mutex_valid as a spin lock -- this may cause
+             * deadlocks in some runtimes.  Applications can fully mitigate this
+             * race by calling wolfSSL_Init() and wolfSSL_Cleanup()
+             * consecutively at startup, uncontended.  This leaves
+             * inits_count_mutex safely initialized, but releases all other
+             * resources.
+             */
+            while ((current_inits_count_mutex_valid = inits_count_mutex_valid) == 0);
+            if (current_inits_count_mutex_valid < 0) {
+                (void)wolfSSL_Atomic_Int_FetchSub(&inits_count_mutex_valid2, 1);
+                return BAD_MUTEX_E;
+            }
```

Comment:
I'm not convinced a spin lock is a good idea to hide in the init function. Since this should be a very rare occurrence, I think this should be treated the same as failing to lock a mutex and returning `BAD_MUTEX_E`.

Reply:
that was indeed the other solution I considered.  I agree that might be better, but we can't use `BAD_MUTEX_E` for the error, because the caller needs to be told that the call is retryable.  (we already return `BAD_MUTEX_E` if `wc_InitMutex()` fails, and that's not retryable.)

with your nudge, I'll switch it to the other solution, and pick or add an appropriate error code that makes it clear by its name that the call should be retried by the application.  effectively the spin locking will move to be the application's responsibility, which is better.